File: plugins/reporting/skills/devexpress-reporting-core/examples/quickstart.vb
' DevExpress XtraReports — Core Runtime API Quickstart (VB.NET)
' Demonstrates: data class, grouped report with XRTable layout, export to PDF

Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports DevExpress.Drawing              ' DXMargins, DXFont, DXFontStyle
Imports DevExpress.Drawing.Printing     ' DXPaperKind
Imports DevExpress.XtraReports.UI
Imports DevExpress.XtraPrinting

' Data model
' Place model classes in a separate file in real projects (see Constraint 10).
Public Class Product
    Public Property Name As String
    Public Property Category As String
    Public Property Price As Decimal
    Public Property Quantity As Integer
End Class

' Report class
Public Class ProductCatalogReport
    Inherits XtraReport

    Public Sub New()
        Me.Name = "ProductCatalogReport"
        Me.PaperKind = DXPaperKind.A4
        Me.Margins = New DXMargins(40, 40, 30, 30)  ' Left, Right, Top, Bottom
        BuildBands()
    End Sub

    Private Sub BuildBands()
        Dim availableWidth As Single = PageWidthF - Margins.Left - Margins.Right

        ' Column proportions (3 : 1 : 1)
        Dim nameW  As Single = availableWidth * 0.6F
        Dim priceW As Single = availableWidth * 0.2F
        Dim qtyW   As Single = availableWidth - nameW - priceW

        Dim boldSmall As New DXFont("Arial", 9F, DXFontStyle.Bold)

        ' Report header
        Dim reportHeader As New ReportHeaderBand()
        Bands.Add(reportHeader)
        reportHeader.HeightF = 50

        Dim titleLabel As New XRLabel()
        titleLabel.Text = "Product Catalog"
        titleLabel.Font = New DXFont("Arial", 16F, DXFontStyle.Bold)
        titleLabel.TextAlignment = TextAlignment.MiddleCenter
        reportHeader.Controls.Add(titleLabel)
        titleLabel.BoundsF = New RectangleF(0, 5, availableWidth, 40)

        ' Group header — sort and display category
        Dim groupHeader As New GroupHeaderBand()
        groupHeader.GroupFields.Add(New GroupField("Category", XRColumnSortOrder.Ascending))
        Bands.Add(groupHeader)
        groupHeader.HeightF = 22

        Dim catLabel As New XRLabel()
        catLabel.Font = boldSmall
        catLabel.ForeColor = Color.DarkBlue
        catLabel.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[Category]"))
        groupHeader.Controls.Add(catLabel)
        catLabel.BoundsF = New RectangleF(0, 2, availableWidth * 0.4F, 18)

        ' Detail band 
        Dim detail As New DetailBand()
        Bands.Add(detail)
        detail.HeightF = 20

        Dim dataTable As New XRTable()
        detail.Controls.Add(dataTable)
        dataTable.BeginInit()
        Dim dataRow As New XRTableRow()
        Dim nameCell  As New XRTableCell() With { .WidthF = nameW }
        Dim priceCell As New XRTableCell() With { .WidthF = priceW, .TextAlignment = TextAlignment.MiddleRight, .TextFormatString = "{0:C2}" }
        Dim qtyCell   As New XRTableCell() With { .WidthF = qtyW,   .TextAlignment = TextAlignment.MiddleRight }
        nameCell.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[Name]"))
        priceCell.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[Price]"))
        qtyCell.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "[Quantity]"))
        dataRow.Cells.Add(nameCell)
        dataRow.Cells.Add(priceCell)
        dataRow.Cells.Add(qtyCell)
        dataTable.Rows.Add(dataRow)
        dataTable.SizeF = New SizeF(availableWidth, 18)
        dataTable.EndInit()

        ' Group footer — subtotal
        Dim groupFooter As New GroupFooterBand()
        Bands.Add(groupFooter)
        groupFooter.HeightF = 22

        Dim subtotalLabel As New XRLabel()
        subtotalLabel.TextAlignment = TextAlignment.MiddleRight
        subtotalLabel.Font = boldSmall
        subtotalLabel.TextFormatString = "Category total: {0:C2}"
        subtotalLabel.Summary = New XRSummary() With { .Running = SummaryRunning.Group }
        subtotalLabel.ExpressionBindings.Add(New ExpressionBinding("BeforePrint", "Text", "sumSum([Price] * [Quantity])"))
        groupFooter.Controls.Add(subtotalLabel)
        subtotalLabel.BoundsF = New RectangleF(0, 2, availableWidth, 18)

        ' Page footer — page number
        Dim pageFooter As New PageFooterBand()
        Bands.Add(pageFooter)
        pageFooter.HeightF = 20

        Dim pageNum As New XRPageInfo()
        pageNum.PageInfo = PageInfo.NumberOfTotal
        pageNum.TextAlignment = TextAlignment.MiddleRight
        pageFooter.Controls.Add(pageNum)
        pageNum.BoundsF = New RectangleF(availableWidth * 0.76F, 2, availableWidth * 0.24F, 16)
    End Sub
End Class

' Usage
Module Program
    Sub Main()
        Dim data As New List(Of Product)()
        data.Add(New Product() With {.Name = "Widget A", .Category = "Tools", .Price = 9.99D, .Quantity = 50})
        data.Add(New Product() With {.Name = "Widget B", .Category = "Tools", .Price = 14.99D, .Quantity = 30})
        data.Add(New Product() With {.Name = "Bolt Pack", .Category = "Hardware", .Price = 4.99D, .Quantity = 200})

        Dim report As New ProductCatalogReport()
        report.DataSource = data

        report.ExportToPdf("catalog.pdf")
        report.ExportToXlsx("catalog.xlsx")
    End Sub
End Module

File: plugins/reporting/skills/devexpress-reporting-core/examples/quickstart.cs
// DevExpress XtraReports — Core Runtime API Quickstart
// Demonstrates: data class, report with header/grouping/detail/footer, export to PDF and XLSX

using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using DevExpress.Drawing;              // DXMargins, DXFont, DXFontStyle
using DevExpress.Drawing.Printing;     // DXPaperKind
using DevExpress.XtraReports.UI;
using DevExpress.XtraReports.Parameters;
using DevExpress.XtraPrinting;

// --- Data model ---
// Place model classes in a separate file (e.g., Model/Product.cs) in real projects.
public class Product
{
    public string Name { get; set; }
    public string Category { get; set; }
    public decimal Price { get; set; }
    public int Quantity { get; set; }
}

// --- Report class ---
public class ProductCatalogReport : XtraReport
{
    public ProductCatalogReport()
    {
        Name = "ProductCatalogReport";
        PaperKind = DXPaperKind.A4;
        Margins = new DXMargins(40, 40, 30, 30);  // Left, Right, Top, Bottom
        ReportUnit = ReportUnit.HundredthsOfAnInch;

        BuildBands();
    }

    private void BuildBands()
    {
        // Derive available width from page dimensions.
        float availableWidth = PageWidthF - Margins.Left - Margins.Right;

        // Column proportions (3 : 1 : 1)
        float nameW  = availableWidth * 0.60f;
        float priceW = availableWidth * 0.20f;
        float qtyW   = availableWidth - nameW - priceW; // remainder avoids floating-point drift

        var boldSmall = new DXFont("Arial", 9f, DXFontStyle.Bold);

        // Report header — title
        var reportHeader = new ReportHeaderBand();
        Bands.Add(reportHeader);
        reportHeader.HeightF = 50;
        var titleLabel = new XRLabel
        {
            Text = "Product Catalog",
            Font = new DXFont("Arial", 16f, DXFontStyle.Bold),
            TextAlignment = TextAlignment.MiddleCenter
        };
        reportHeader.Controls.Add(titleLabel);
        titleLabel.BoundsF = new RectangleF(0, 0, availableWidth, 50);

        // Page header — column labels
        var pageHeader = new PageHeaderBand();
        Bands.Add(pageHeader);
        pageHeader.HeightF = 22;
        var headerTable = new XRTable();
        pageHeader.Controls.Add(headerTable);
        headerTable.SizeF = new SizeF(availableWidth, 22);
        headerTable.BeginInit();
        var headerRow = new XRTableRow();        
        headerTable.Rows.Add(headerRow);
        headerRow.Cells.Add(new XRTableCell { Text = "Product", WidthF = nameW,  Font = boldSmall });
        headerRow.Cells.Add(new XRTableCell { Text = "Price",   WidthF = priceW, Font = boldSmall, TextAlignment = TextAlignment.MiddleRight });
        headerRow.Cells.Add(new XRTableCell { Text = "Qty",     WidthF = qtyW,   Font = boldSmall, TextAlignment = TextAlignment.MiddleRight });
        headerTable.EndInit();

        // Group header — category name
        var groupHeader = new GroupHeaderBand
        {
            GroupFields = { new GroupField("Category", XRColumnSortOrder.Ascending) }
        };
        Bands.Add(groupHeader);
        groupHeader.HeightF = 22;
        var catLabel = new XRLabel
        {
            Font = boldSmall,
            ForeColor = Color.DarkBlue
        };
        catLabel.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[Category]"));
        groupHeader.Controls.Add(catLabel);
        catLabel.BoundsF = new RectangleF(0, 0, availableWidth * 0.40f, 22);

        // Detail band — multi-column table
        var detail = new DetailBand();
        Bands.Add(detail);
        detail.HeightF = 20;
        var dataTable = new XRTable();
        detail.Controls.Add(dataTable);
        dataTable.SizeF = new SizeF(availableWidth, 20);
        dataTable.BeginInit();
        var dataRow = new XRTableRow();
        dataTable.Rows.Add(dataRow);
        var nameCell  = new XRTableCell { WidthF = nameW };
        var priceCell = new XRTableCell { WidthF = priceW, TextAlignment = TextAlignment.MiddleRight };
        var qtyCell   = new XRTableCell { WidthF = qtyW,   TextAlignment = TextAlignment.MiddleRight };
        nameCell.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[Name]"));
        priceCell.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[Price]"));
        priceCell.TextFormatString = "{0:C2}";
        qtyCell.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "[Quantity]"));
        dataRow.Cells.Add(nameCell);
        dataRow.Cells.Add(priceCell);
        dataRow.Cells.Add(qtyCell);
        dataTable.EndInit();

        // Group footer — subtotal per category
        var groupFooter = new GroupFooterBand();
        Bands.Add(groupFooter);
        groupFooter.HeightF = 22;
        var subtotalLabel = new XRLabel
        {
            TextAlignment = TextAlignment.MiddleRight,
            Font = boldSmall,
            TextFormatString = "Category total: {0:C2}",
            Summary = new XRSummary { Running = SummaryRunning.Group }
        };
        subtotalLabel.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "sumSum([Price] * [Quantity])"));
        groupFooter.Controls.Add(subtotalLabel);
        subtotalLabel.BoundsF = new RectangleF(0, 0, availableWidth, 22);

        // Report footer — grand total
        var reportFooter = new ReportFooterBand();
        Bands.Add(reportFooter);
        reportFooter.HeightF = 30;
        var grandTotalLabel = new XRLabel
        {
            TextAlignment = TextAlignment.MiddleRight,
            Font = new DXFont("Arial", 10f, DXFontStyle.Bold),
            ForeColor = Color.DarkGreen,
            TextFormatString = "Grand total: {0:C2}",
            Summary = new XRSummary { Running = SummaryRunning.Report }
        };
        grandTotalLabel.ExpressionBindings.Add(new ExpressionBinding("BeforePrint", "Text", "sumSum([Price] * [Quantity])"));
        reportFooter.Controls.Add(grandTotalLabel);
        grandTotalLabel.BoundsF = new RectangleF(0, 0, availableWidth, 30);

        // Page footer — page number    
        var pageFooter = new PageFooterBand();
        Bands.Add(pageFooter);
        pageFooter.HeightF = 20;
        var pageNum = new XRPageInfo
        {
            PageInfo = PageInfo.NumberOfTotal,
            TextAlignment = TextAlignment.MiddleRight
        };
        pageFooter.Controls.Add(pageNum);
        pageNum.BoundsF = new RectangleF(availableWidth * 0.76f, 0, availableWidth * 0.24f, 20);
    }
}

// --- Usage ---
public static class Program
{
    public static void Main()
    {
        var data = new List<Product>
        {
            new Product { Name = "Widget A", Category = "Tools", Price = 9.99m, Quantity = 50 },
            new Product { Name = "Widget B", Category = "Tools", Price = 14.99m, Quantity = 30 },
            new Product { Name = "Bolt Pack", Category = "Hardware", Price = 4.99m, Quantity = 200 },
            new Product { Name = "Nut Pack", Category = "Hardware", Price = 3.49m, Quantity = 150 },
        };

        var report = new ProductCatalogReport { DataSource = data };

        // Export to PDF
        report.ExportToPdf("catalog.pdf");

        // Export to XLSX with options
        var xlsxOptions = new XlsxExportOptions
        {
            SheetName = "Products",
            TextExportMode = TextExportMode.Value
        };
        report.ExportToXlsx("catalog.xlsx", xlsxOptions);

        // Export to stream (for web/API use)
        using var ms = new MemoryStream();
        report.ExportToPdf(ms);
        // ms is ready for return File(ms.ToArray(), "application/pdf", "catalog.pdf")
    }
}
